rust/trace/src/fixed_lifo_deque.rs

```diff
@@ -0,0 +1,374 @@
+        self.storage.front()
+    }
+
+    #[inline]
+    pub fn front_mut(&mut self) -> Option<&mut T> {
+        self.storage.front_mut()
+    }
+
+    #[inline]
+    pub fn back(&self) -> Option<&T> {
+        self.storage.back()
+    }
+
+    #[inline]
+    pub fn back_mut(&mut self) -> Option<&mut T> {
+        self.storage.back_mut()
+    }
+
+    #[inline]
+    fn drop_excess_for_inserting(&mut self, n_to_be_inserted: usize) {
+        if self.storage.len() + n_to_be_inserted > self.limit {
+            let overflow = cmp::min(self.storage.len(), self.storage.len() + n_to_be_inserted - self.limit);
+            self.storage.drain(..overflow);
+        }
+    }
+
+    #[inline]
+    pub fn pop_front(&mut self) -> Option<T> {
+        self.storage.pop_front()
+    }
+
+    pub fn push_front(&mut self, value: T) {
+        self.drop_excess_for_inserting(1);
+        self.storage.push_front(value);
+        self.drop_excess_for_inserting(0);
+    }
+
+    pub fn push_back(&mut self, value: T) {
+        self.drop_excess_for_inserting(1);
+        self.storage.push_back(value);
+        self.drop_excess_for_inserting(0);
+    }
+
+    #[inline]
+    pub fn pop_back(&mut self) -> Option<T> {
+        self.storage.pop_back()
+    }
+
+    #[inline]
+    pub fn swap_remove_back(&mut self, index: usize) -> Option<T> {
+        self.storage.swap_remove_back(index)
+    }
+
+    #[inline]
+    pub fn swap_remove_front(&mut self, index: usize) -> Option<T> {
+        self.storage.swap_remove_front(index)
+    }
+    
+    pub fn insert(&mut self, index: usize, value: T) {
+        self.drop_excess_for_inserting(1);
+        self.storage.insert(index, value);
+        self.drop_excess_for_inserting(0);
+    }
+
+    #[inline]
+    pub fn remove(&mut self, index: usize) -> Option<T> {
+        self.storage.remove(index)
+    }
+    
+    pub fn split_off(&mut self, at: usize) -> FixedLifoDeque<T> {
+        FixedLifoDeque {
+            storage: self.storage.split_off(at),
+            limit: self.limit
+        }
+    }
+
+    pub fn append(&mut self, other: &mut VecDeque<T>) {
+        self.drop_excess_for_inserting(other.len());
+        self.storage.append(other);
+        self.drop_excess_for_inserting(0);
+    }
+
+    #[inline]
+    pub fn retain<F>(&mut self, f: F)
+        where F: FnMut(&T) -> bool
+    {
+        self.storage.retain(f);
+    }
+}
+
+impl<T: Clone> FixedLifoDeque<T> {
+    /// Resizes a fixed queue.  This doesn't change the limit so the resize is
+    /// capped to the limit.  Additionally, resizing drops the elements from the front
+    /// unlike with a regular VecDeque.
+    pub fn resize(&mut self, new_len: usize, value: T) {
+        if new_len < self.len() {
+            let to_drop = self.len() - new_len;
+            self.storage.drain(..to_drop);
+        } else {
+            self.storage.resize(cmp::min(self.limit, new_len), value);
+        }
+    }
+}
+
+impl<A: PartialEq> PartialEq for FixedLifoDeque<A> {
```

Comment:
I'll be honest.  I just copied the pattern from VecDeque.  However, I think it makes sense because I don't think the limit should change anything with respect to comparison.  It would certainly be a very surprising result to me.  It's like capacity.  One thought I have, looking at this again, is to make the limit & capacity identical since I think it is always.  That might simplify this on several fronts.  Thoughts?

Reply:
I'm not sure what that would buy us, unless I'm misunderstanding. I believe capacity is just the size of the current allocation, not a ceiling.

Reply:
Well if you could guarantee that the capacity was exactly what you ever asked for, then it could also act as the limit since for this kind of data structure it wouldn't make sense to ever request a larger capacity (& it never grows).  What it saves you is that you can just derive the default traits since your only member would be a VecDeque which already implements everything correctly.  Thinking about it some more, I'm not sure that the API ever guarantees anything about the capacity beyond a lower bound so in theory you may get back more.  That being said, maybe it doesn't matter?  If you ask for a limit of 500 but the allocator returned enough for 600, does it matter?  Maybe if you cared about FixedLifoDeque as a general container...  Alternatively, maybe something like https://github.com/mcarton/rust-derivative/blob/master/doc/cmp.md?

Reply:
I think it's fine as written, was more just curious. I'd have probably derived and then defensively argued that limit length is in fact _fundamental_ to the concept of equality for this type. 😉